File: src/main/java/pl/droidsonroids/gif/GifViewSavedState.java
package pl.droidsonroids.gif;

import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.view.View;

class GifViewSavedState extends View.BaseSavedState {

    final long[][] mStates;

    GifViewSavedState(Parcelable superState, Drawable... drawables) {
        super(superState);
        mStates = new long[drawables.length][];
        for (int i = 0; i < drawables.length; i++) {
            Drawable drawable = drawables[i];
            if (drawable instanceof GifDrawable) {
                mStates[i] = ((GifDrawable) drawable).mNativeInfoHandle.getSavedState();
            } else {
                mStates[i] = null;
            }
        }
    }

    private GifViewSavedState(Parcel in) {
        super(in);
        mStates = new long[in.readInt()][];
        for (int i=0;i<mStates.length;i++)
            mStates[i]=in.createLongArray();
    }

    public GifViewSavedState(Parcelable superState, long[] savedState) {
        super(superState);
        mStates = new long[1][];
        mStates[0] = savedState;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeInt(mStates.length);
        for (long[] mState : mStates)
            dest.writeLongArray(mState);
    }

    public static final Creator<GifViewSavedState> CREATOR = new Creator<GifViewSavedState>() {

        public GifViewSavedState createFromParcel(Parcel in) {
            return new GifViewSavedState(in);
        }

        public GifViewSavedState[] newArray(int size) {
            return new GifViewSavedState[size];
        }
    };

    void restoreState(Drawable drawable, int i) {
        if (mStates[i] != null && drawable instanceof GifDrawable) {
            final GifDrawable gifDrawable = (GifDrawable) drawable;
            gifDrawable.startAnimation(gifDrawable.mNativeInfoHandle.restoreSavedState(mStates[i], gifDrawable.mBuffer));
        }
    }
}

File: android-gif-drawable/src/main/java/pl/droidsonroids/gif/GifViewSavedState.kt
package pl.droidsonroids.gif

import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import android.view.View

internal class GifViewSavedState : View.BaseSavedState {
    val mStates: Array<LongArray?>

    constructor(superState: Parcelable?, vararg drawables: Drawable?) : super(superState) {
        mStates = arrayOfNulls(drawables.size)
        for (i in drawables.indices) {
            val drawable = drawables[i]
            if (drawable is GifDrawable) {
                mStates[i] = drawable.mNativeInfoHandle.savedState
            } else {
                mStates[i] = null
            }
        }
    }

    private constructor(`in`: Parcel) : super(`in`) {
        mStates = arrayOfNulls(`in`.readInt())
        for (i in mStates.indices) mStates[i] = `in`.createLongArray()
    }

    constructor(superState: Parcelable?, savedState: LongArray?) : super(superState) {
        mStates = arrayOfNulls(1)
        mStates[0] = savedState
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        super.writeToParcel(dest, flags)
        dest.writeInt(mStates.size)
        for (mState in mStates) dest.writeLongArray(mState)
    }

    fun restoreState(drawable: Drawable?, i: Int) {
        if (mStates[i] != null && drawable is GifDrawable) {
            drawable.startAnimation(
                drawable.mNativeInfoHandle.restoreSavedState(
                    mStates[i]!!,
                    drawable.mBuffer
                ).toLong()
            )
        }
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Creator<GifViewSavedState> {
        override fun createFromParcel(parcel: Parcel): GifViewSavedState {
            return GifViewSavedState(parcel)
        }

        override fun newArray(size: Int): Array<GifViewSavedState?> {
            return arrayOfNulls(size)
        }
    }
}